### ISSUE TITLE
YAPF crashes on assignment expressions, and context does not include file name

### ISSUE THREAD

**Issue by agentydragon** [google/yapf#867]
I get this error running on a file of mine:

```
Traceback (most recent call last):
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 115, in ParseCodeToTree
    tree = parser_driver.parse_string(code, debug=False)
  File "/usr/lib/python3.8/lib2to3/pgen2/driver.py", line 103, in parse_string
    return self.parse_tokens(tokens, debug)
  File "/usr/lib/python3.8/lib2to3/pgen2/driver.py", line 71, in parse_tokens
    if p.addtoken(type, value, (prefix, start)):
  File "/usr/lib/python3.8/lib2to3/pgen2/parse.py", line 162, in addtoken
    raise ParseError("bad input", type, value, context)
lib2to3.pgen2.parse.ParseError: bad input: type=11, value=':', context=(' ', (478, 18))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/agentydragon/.local/bin/yapf", line 8, in <module>
    sys.exit(run_main())
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/__init__.py", line 365, in run_main
    sys.exit(main(sys.argv))
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/__init__.py", line 125, in main
    changed = FormatFiles(
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/__init__.py", line 203, in FormatFiles
    changed |= _FormatFile(filename, lines, style_config, no_local_style,
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/__init__.py", line 226, in _FormatFile
    reformatted_code, encoding, has_change = yapf_api.FormatFile(
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/yapflib/yapf_api.py", line 94, in FormatFile
    reformatted_source, changed = FormatCode(
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/yapflib/yapf_api.py", line 147, in FormatCode
    tree = pytree_utils.ParseCodeToTree(unformatted_source)
  File "/home/agentydragon/.local/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 121, in ParseCodeToTree
    tree = parser_driver.parse_string(code, debug=False)
  File "/usr/lib/python3.8/lib2to3/pgen2/driver.py", line 103, in parse_string
    return self.parse_tokens(tokens, debug)
  File "/usr/lib/python3.8/lib2to3/pgen2/driver.py", line 71, in parse_tokens
    if p.addtoken(type, value, (prefix, start)):
  File "/usr/lib/python3.8/lib2to3/pgen2/parse.py", line 162, in addtoken
    raise ParseError("bad input", type, value, context)
lib2to3.pgen2.parse.ParseError: bad input: type=11, value=':', context=(' ', (478, 18))
```

The code around line 478 looks like this:

```
class SomeClass:
    ...
    def _get_products(self) -> List[planner_pb2.Product]:
        return [
            p for product in self._products
            if (p := self._get_product(product)) # <-- this line
        ]
```

As far as I know this is a perfectly valid assignment expression (https://www.python.org/dev/peps/pep-0572/). Would be nice if the exception would include the filename (otherwise I'm left guessing which file in my project is doing this), and also would be nice to support those.

**Comment by kamahen** [comment]
This is a known problem: Issue #772

**Comment by agentydragon** [comment]
Ah, I agree, I missed the canonical issue for python 3.8. would be still
nice to fix the stack trace so that it contains the filename

**Comment by kamahen** [comment]
Looking at the code (and OP's stack trace), we start with `ParseError: bad input: type=11, value=':', context=(' ', (478, 18))` which should be caught in yapf_api.Formatcode and have the filename added:
```
  try:
    tree = pytree_utils.ParseCodeToTree(unformatted_source)
  except parse.ParseError as e:
    e.msg = filename + ': ' + e.msg
    raise
```
However, that doesn't seem to show in the traceback. 

I know that there's some magic for handling `SyntaxError` at the top level -- maybe something similar needs to be done here?

**Comment by bwendling** [comment]
Closed with 7c408b9d7750292760ed255f744211d1ef535668.
